typing/ctype.ml

```diff
@@ -328,27 +328,71 @@ let none = newty (Ttuple [])                (* Clearly ill-formed type *)
 (**** information for [Typecore.unify_pat_*] ****)
 
 module Pattern_env : sig
+  type envop
   type t = private
     { mutable env : Env.t;
+      mutable op_list : envop list;
       equations_scope : int;
       in_counterexample : bool; }
   val make: Env.t -> equations_scope:int -> in_counterexample:bool -> t
   val copy: ?equations_scope:int -> t -> t
+  val enter_type: scope:int -> label -> type_declaration -> t -> Ident.t
+  val add_local_constraint: Path.t -> type_declaration -> t -> unit
+  val with_mty: t -> Ident.Unscoped.t -> module_type -> (unit -> 'a) -> 'a
   val set_env: t -> Env.t -> unit
 end = struct
+  type envop =
+    | Enter_type of int * label * type_declaration
+    | Local_constraint of Path.t * type_declaration
+
   type t =
     { mutable env : Env.t;
+      mutable op_list : envop list;
       equations_scope : int;
       in_counterexample : bool; }
   let make env ~equations_scope ~in_counterexample =
     { env;
+      op_list = [];
       equations_scope;
       in_counterexample; }
   let copy ?equations_scope penv =
     let equations_scope =
       match equations_scope with None -> penv.equations_scope | Some s -> s in
     { penv with equations_scope }
-  let set_env penv env = penv.env <- env
+
+  let enter_type ~scope lbl decl penv =
+    let (id, new_env) = Env.enter_type ~scope lbl decl penv.env in
+    penv.env <- new_env;
+    penv.op_list <- (Enter_type (scope, lbl, decl)) :: penv.op_list;
+    id
+
+  let add_local_constraint source dest penv =
+    let new_env = Env.add_local_constraint source dest penv.env in
+    penv.env <- new_env;
+    penv.op_list <- (Local_constraint (source, dest)) :: penv.op_list
+
+  let do_op penv = function
+    | Enter_type (_scope, _lbl, _decl) ->
+      assert false
```

Comment:
This assert false is reachable in pattern mode when the typechecker reifies type variables:
```ocaml

module R = struct
  type 'a n
end

module type S = sig
  type 'a t = X
  type u
  type v
end

let f ( (module X:S) -> (int * 'a) R.n, (module X:S) -> (float * 'b) R.n) Type.eq -> _ = 
  function
  | Type.Equal -> ()
```

Reply:
In addition, we believe that there should be documentation that explains the design that you have chosen: to record the operations performed to be able to reply them in the "parent" environment. It is a bit opaque as-is for people that read this part of the code without being familiar with the Tfunctor logic -- simply explaining *when* those facilities are used would already be an improvement.